pkg/services/accesscontrol/ossaccesscontrol/permissions_services.go

```diff
@@ -186,6 +190,83 @@ func ProvideDashboardPermissions(
 	return &DashboardPermissionsService{srv}, nil
 }
 
+type LibraryElementPermissionsService struct {
+	*resourcepermissions.Service
+}
+
+var LibraryElementViewActions = []string{libraryelements.ActionLibraryPanelsRead}
+var LibraryElementEditActions = append(LibraryElementViewActions, []string{libraryelements.ActionLibraryPanelsWrite, libraryelements.ActionLibraryPanelsDelete}...)
+var LibraryElementAdminActions = LibraryElementEditActions
+
+func ProvideLibraryElementPermissions(
```

Comment:
added to LibraryElementService in libraryelements.go:
```go
func ProvideService(cfg *setting.Cfg, sqlStore db.DB, routeRegister routing.RouteRegister, folderService folder.Service, features featuremgmt.FeatureToggles, ac accesscontrol.AccessControl, libraryElementPermissionsService accesscontrol.LibraryElementPermissionsService) *LibraryElementService {
	l := &LibraryElementService{
		Cfg:                       cfg,
		SQLStore:                  sqlStore,
		RouteRegister:             routeRegister,
		folderService:             folderService,
		log:                       log.New("library-elements"),
		features:                  features,
		AccessControl:             ac,
		libraryElementPermissions: libraryElementPermissionsService,
	}

	l.registerAPIEndpoints()
	ac.RegisterScopeAttributeResolver(LibraryPanelUIDScopeResolver(l, l.folderService))

	return l
}
```

and then used as part of `updatePermissions` in database.go (second last line)

```go
func (l *LibraryElementService) updatePermissions(c context.Context, model *model.LibraryElementDTO, signedInUser identity.Requester) error {
	inFolder := model.FolderID > 0
	var permissions []accesscontrol.SetResourcePermissionCommand
	namespaceID, userIDstr := signedInUser.GetNamespacedID()
	userID, err := identity.IntIdentifier(namespaceID, userIDstr)
	if err != nil {
		return err
	}

	if namespaceID == identity.NamespaceUser {
		permissions = append(permissions, accesscontrol.SetResourcePermissionCommand{
			UserID: userID, Permission: dashboards.PERMISSION_ADMIN.String(),
		})
	}

	if !inFolder {
		permissions = append(permissions, []accesscontrol.SetResourcePermissionCommand{
			{BuiltinRole: string(org.RoleEditor), Permission: dashboards.PERMISSION_EDIT.String()},
			{BuiltinRole: string(org.RoleViewer), Permission: dashboards.PERMISSION_VIEW.String()},
		}...)
	}

	_, err = l.libraryElementPermissions.SetPermissions(c, model.OrgID, model.UID, permissions...)
	return err
}
```

Reply:
Thanks for linking the usage. The way we've used the permission service framework so far is for resources that have a frontend component that allows setting permissions on them through the UI (like [this tab](https://ops.grafana-ops.net/d/d8e4202e-6737-4ec2-b62d-c15de841663a/api-key-deprecation?orgId=1&editview=permissions) for dashboards). So I feel like it might be an overkill for library elements.

Reply:
That seems reasonable